contracts/registry/Registry.sol

```diff
@@ -469,4 +411,135 @@ contract Registry is Governance {
+        // Get the relase version for this specific vault.
+        uint256 releaseTarget = ReleaseRegistry(releaseRegistry).releaseTargets(
+            IVault(_vault).api_version()
+        );
+
+        address[] memory _vaults = _endorsedVaults[asset];
+        require(_vaults[_index] == _vault, "wrong index");
+
+        // Set the last index to the spot we are removing.
+        _vaults[_index] = _vaults[_vaults.length - 1];
+
+        // Set to storage.
+        _endorsedVaults[asset] = _vaults;
+
+        // Pop the last item off the array.
+        _endorsedVaults[asset].pop();
+
+        // Reset the info config.
+        info[_vault] = Info({
+            asset: address(0),
+            releaseVersion: 0,
+            deploymentTimeStamp: 0,
+            tag: ""
+        });
+
+        // Emit the event.
+        emit RemovedVault(_vault, asset, releaseTarget);
+    }
+
+    /**
+     * @notice Remove a `_strategy` at a specific `_index`.
+     * @dev Can be used as a efficent way to remove a strategy
+     * to not have to iterate over the full array.
+     *
+     * NOTE: This will not remove the asset from the `assets` array
+     * if it is no longer in use and will have to be done manually.
+     *
+     * @param _strategy Address of the strategy to remove.
+     * @param _index Index in the `endorsedStrategies` array `_strategy` sits at.
+     */
+    function removeStrategy(
+        address _strategy,
+        uint256 _index
+    ) external onlyGovernance {
+        require(info[_strategy].asset != address(0), "!endorsed");
+
+        // Get the asset the vault is using.
+        address asset = IStrategy(_strategy).asset();
+        // Get the relase version for this specific vault.
+        uint256 releaseTarget = ReleaseRegistry(releaseRegistry).releaseTargets(
+            IStrategy(_strategy).apiVersion()
+        );
+
+        address[] memory _strategies = _endorsedStrategies[asset];
+        require(_strategies[_index] == _strategy, "wrong index");
+
+        // Set the last index to the spot we are removing.
+        _strategies[_index] = _strategies[_strategies.length - 1];
+
+        // Set to storage.
+        _endorsedStrategies[asset] = _strategies;
+
+        // Pop the last item off the array.
+        _endorsedStrategies[asset].pop();
+
+        // Reset the info config.
+        info[_strategy] = Info({
+            asset: address(0),
+            releaseVersion: 0,
+            deploymentTimeStamp: 0,
+            tag: ""
+        });
+
+        // Emit the event.
+        emit RemovedStrategy(_strategy, asset, releaseTarget);
+    }
+
+    /**
+     * @notice Removes a specific `_asset` at `_index` from `assets`.
+     * @dev Can be used if an asset is no longer in use after a vault or
+     * strategy has also been removed.
+     *
+     * @param _asset The asset to remove from the array.
+     * @param _index The index it sits at.
+     */
+    function removeAsset(
```

Comment:
currently, governance has to remember to call this.
is there a good way to ensure this gets called automatically (when appropriate) when/if all strats/vaults get removed for this asset?

Reply:
Yes. Thats how i had made it orginally.
https://github.com/yearn/vault-periphery/blob/5f48dad2ae2c5e528827a2aefd189d43b598250f/contracts/registry/Registry.sol#L438-L447

https://github.com/yearn/vault-periphery/blob/5f48dad2ae2c5e528827a2aefd189d43b598250f/contracts/registry/Registry.sol#L523-L539

Problem is that auto removing it has to iterate over the entire assets array. Which could possibly be 30-40 + items.

So it seemed better to not force the gas cost of that and allow governance to just remove it manually if needed like the strategies and vaults.

I dont think it should cause any real problems for it to be left in even if not in use.